src/Compilers/CSharp/Portable/Binder/UnionMatchingRewriter.cs

```diff
@@ -0,0 +1,449 @@
+        }
+
+        public override BoundNode? VisitPositionalSubpattern(BoundPositionalSubpattern node)
+        {
+            Symbol? symbol = node.Symbol;
+            BoundPattern pattern = RewritePatternWithUnionMatchingToPropertyPattern((BoundPattern)this.Visit(node.Pattern));
+            return node.Update(symbol, pattern);
+        }
+
+        public override BoundNode? VisitPropertySubpattern(BoundPropertySubpattern node)
+        {
+            BoundPropertySubpatternMember? member = node.Member;
+            BoundPattern pattern = RewritePatternWithUnionMatchingToPropertyPattern((BoundPattern)this.Visit(node.Pattern));
+            return node.Update(member, node.IsLengthOrCount, pattern);
+        }
+
+        public override BoundNode? VisitBinaryPattern(BoundBinaryPattern node)
+        {
+            var binaryPatternStack = ArrayBuilder<BoundBinaryPattern>.GetInstance();
+            BoundBinaryPattern? currentNode = node;
+
+            do
+            {
+                binaryPatternStack.Push(currentNode);
+                currentNode = currentNode.Left as BoundBinaryPattern;
+            } while (currentNode != null);
+
+            Debug.Assert(binaryPatternStack.Count > 0);
+
+            var binaryPattern = binaryPatternStack.Pop();
+            BoundPattern result = (BoundPattern)Visit(binaryPattern.Left);
+#if DEBUG
+            var narrowedTypeCandidates = ArrayBuilder<TypeSymbol>.GetInstance(2);
+
+            if (result is BoundPatternWithUnionMatching unionPattern)
+            {
+                narrowedTypeCandidates.Add(getDisjunctionType(unionPattern));
+            }
+            else
+            {
+                Binder.CollectDisjunctionTypes(result, narrowedTypeCandidates, hasUnionMatching: false);
+            }
+#endif
+            do
+            {
+                result = rewriteBinaryPattern(
+                    this,
+                    result,
+                    binaryPattern
+#if DEBUG
+                    , narrowedTypeCandidates
+#endif
+                    );
+            } while (binaryPatternStack.TryPop(out binaryPattern));
+
+            binaryPatternStack.Free();
+#if DEBUG
+            narrowedTypeCandidates.Free();
+#endif
+            return result;
+
+            static BoundPattern rewriteBinaryPattern(
+                UnionMatchingRewriter rewriter,
+                BoundPattern preboundLeft,
+                BoundBinaryPattern node
+#if DEBUG
+                , ArrayBuilder<TypeSymbol> narrowedTypeCandidates
+#endif
+                )
+            {
+                if (node.Disjunction)
+                {
+                    preboundLeft = RewritePatternWithUnionMatchingToPropertyPattern(preboundLeft);
+                    var right = RewritePatternWithUnionMatchingToPropertyPattern((BoundPattern)rewriter.Visit(node.Right));
+
+#if DEBUG
+                    // Compute the common type. This algorithm is quadratic, but disjunctive patterns are unlikely to be huge
```

Comment:
Perhaps a comment here about keeping in sync as well? It might also be worth leaving a quick sentence that this is just here to make sure that the type narrowing code lines up, I was confused when I first read it and the explanation was quite helpful.